docs/adr/2022-07-22_019-reference-scripts.md

```diff
@@ -0,0 +1,61 @@
+---
+slug: 19
+title: | 
+  19. Use of reference scripts
+authors: []
+tags: [Proposed]
+---
+
+## Status
+
+Proposed
+
+## Context
+
+* In the desire to make Hydra transactions smaller and cheaper (at the time of writing any abort tx was too big), we want to use the **reference script** and **reference input** features of the upcoming `Babbage` ledger era. See the [babbage ledger spec](https://hydra.iohk.io/build/16861604/download/1/babbage-changes.pdf), [CIP-31](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0031) and [CIP-33](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0033) for details.
+
+* With these features we do not need to (re-)include scripts in each transaction.
+
+* The CIPs do not specify how reference scripts are to be managed and we can see at least two options:
+  1. Add them as outputs to the `init` transaction or prior that as part of each Hydra Head instance
+  2. Post them out-of-band, separate to individual Head instances
+
+* Ownership of the outputs holding the scripts is to be considered. If these "reference outputs" are spent, they cannot be referred to anymore. This would mean all heads referring to them can be denied of service (DoS).
+
+* Each head will need to refer to the correct version of the hydra scripts. That is, consistent with the script hashes known to the `hydra-node`.
+  + This is also related to the problem of managing script versions & updates.
+  + Right now, the `hydra-node` is compiled against `hydra-plutus` to access compiled script content and hashes.
+
+* The general trade-off is: instead of paying ADA fees for scripts adding to the transaction size in _each_ transaction, ADA deposits will need to be put down to have scripts be part of the UTxO set in the ledger _once_.
+
+## Decision
+
+* Publish outputs holding Hydra scripts out-of-band (option 2), because
+  + All scripts would not fit into the `init` transaction directly, we would need to post multiple.
+  + Costs (deposits) would need to be payed for each head instance.
+  + Although, heads using option 1 would not be exposed to DoS 
+
+* We will publish scripts on release of the `hydra-node`, or more specifically of the `hydra-plutus` package.
+
+* The scripts are stored at outputs addressed to some **unspendable** `v_publish` validator.
+  + This is to avoid DoS risk and unnecessariy centralization with option 2
+
+## Consequences
+
+* We need a process and/or tool to publish `hydra-plutus` scripts and need to pay the deposits.
+  + Any other party could to the same, this does not lead to centralization.
+
+* The `hydra-node` would be need to know the `TxIn`s of the "right" published scripts. In the simplest case we would just make this configurable.
```

Comment:
I'd park that as an "implementation details" (the `TxIn` bit). As we discussed, we need to provide the Hydra node with the output reference corresponding to each scripts needed as reference. This can be inferred from a `TxIn` + state query (as we currently do), or could be specified as a configuration file / command-line options fully.

Reply:
Are you saying that it's okay to leave it like this and how the `TxIn` gets inferred is an implementation detail? i.e. not enumerating all but doing what we do right now is an optimization on convenience == implementation detail not covered by the adr

Reply:
I mean that mentioning `TxIn` is an implementation detail. The ADR should simply say that a reference to the published scripts must be communicated to the Hydra node. An address, a TxId, a TxIn ... doesn't matter much.

Reply:
Perhaps I am just neat picking.

Reply:
Ah well. Ultimately the reference needs to be available as a `TxIn` eventually, so I think that formulation is the common denominator all specific implementations need to fulfill. :bike: :house: